File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/for.go
package statement

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/ast_types"
)

type For struct {
	StatementBase
	Initialization ast.Statement
	Condition      ast.Expression
	Post           ast.Statement
	Body           ast.Statement
}

func (f *For) Validate(c ast.Context) {
	ctx := c.NewContext()
	if f.Initialization != nil {
		f.Initialization.Validate(ctx)
	}
	if f.Condition != nil {
		conditionCtx := ctx.NewContext()
		f.Condition.Validate(conditionCtx, ast_types.TypeBool)
		if f.Condition.Type() != nil && !f.Condition.Type().Equal(ast_types.TypeBool) {
			c.Error(f.Condition.GetPosition(), "expect bool type condition")
		}
	}
	if f.Post != nil {
		postCtx := ctx.NewContext()
		f.Post.Validate(postCtx)
	}
	if f.Body != nil {
		bodyCtx := ctx.NewContext()
		f.Body.Validate(bodyCtx)
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/For.java
package com.github.panda_io.micro_panda.ast.statement;

import com.github.panda_io.micro_panda.ast.Context;
import com.github.panda_io.micro_panda.ast.type.Type;

public class For extends Statement {
    public Statement initialization;
    public com.github.panda_io.micro_panda.ast.expression.Expression condition;
    public Statement post;
    public Statement body;

    public void validate(Context context) {
        Context ctx = context.newContext();
        if (this.initialization != null) {
            this.initialization.validate(context);
        }
        if (this.condition != null) {
            Context conditionCtx = ctx.newContext();
            this.condition.validate(conditionCtx, Type.bool);
            if (this.condition.getType() != null && !this.condition.getType().equal(Type.bool)) {
                context.addError(this.condition.getOffset(), "expect bool type condition");
            }
        }
        if (this.post != null) {
            Context postCtx = ctx.newContext();
            this.post.validate(postCtx);
        }
        if (this.body != null) {
            Context bodyCtx = ctx.newContext();
            this.body.validate(bodyCtx);
        }
    }
}
